toolchain/check/context.cpp

```diff
@@ -187,6 +187,82 @@ auto Context::DiagnoseNameNotFound(SemIRLoc loc, SemIR::NameId name_id)
   emitter_->Emit(loc, NameNotFound, name_id);
 }
 
+auto Context::DiagnoseInvalidAccess(
+    Parse::NodeId node_id, SemIR::InstId member_id, SemIR::NameId name_id,
+    SemIR::AccessKind access_kind, bool is_parent_access,
+    std::optional<AccessInfo> access_info) -> void {
+  if (!access_info) {
+    return;
+  }
+
+  auto inst = insts().Get(access_info->base_id);
+  auto inst_type_id = inst.type_id();
+
+  if (auto class_type = types().TryGetAs<SemIR::ClassType>(inst_type_id)) {
+    auto class_inst = classes().Get(class_type->class_id);
+    auto is_access_from_instance_method = access_info->allow_protected_access &&
+                                          access_info->allow_private_access;
+
+    CARBON_DCHECK(class_inst.is_defined());
+
+    if (access_kind == SemIR::AccessKind::Protected) {
+      if (!is_access_from_instance_method) {
+        CARBON_DIAGNOSTIC(ClassProtectedMemberAccess, Error,
+                          "Cannot access protected member `{0}` of type `{1}`",
+                          SemIR::NameId, SemIR::TypeId);
+        CARBON_DIAGNOSTIC(ClassProtectedMemberDefinition, Note,
+                          "Protected member `{0}` defined here.`",
+                          SemIR::NameId);
+
+        emitter()
+            .Build(node_id, ClassProtectedMemberAccess, name_id, inst_type_id)
+            .Note(member_id, ClassProtectedMemberDefinition, name_id)
+            .Emit();
+        return;
+      }
+    }
+
+    if (access_kind == SemIR::AccessKind::Private) {
+      CARBON_DIAGNOSTIC(ClassPrivateMemberDefinition, Note,
+                        "Private member `{0}` defined here.`", SemIR::NameId);
+
+      if (!is_access_from_instance_method) {
+        CARBON_DIAGNOSTIC(ClassPrivateMemberAccess, Error,
+                          "Cannot access private member `{0}` of type `{1}`",
+                          SemIR::NameId, SemIR::TypeId);
+        emitter()
+            .Build(node_id, ClassPrivateMemberAccess, name_id, inst_type_id)
+            .Note(member_id, ClassPrivateMemberDefinition, name_id)
+            .Emit();
+        return;
+      }
```

Comment:
These two diagnostics are very similar. Can you make "private" versus "protected" a parameter to a single diagnostic that handles both cases?